feat(kinput): redesign kinput

kaiarrowood opened the pull request:
### Summary
Update `KInput` to new metalab designs.

#### Changes made:
* Add `size` property that can take values `small`, `medium` (default), `large`
* Add border updates on `hover` and `focus`
* Add `--KInputHover` theming var for hover border / label color
* Update snaps / tests

Figma: https://www.figma.com/file/sfSZhMfpzCzddEhd8tdk9s/Kong_Design-library?node-id=209%3A1429

![image](https://user-images.githubusercontent.com/67973710/134416945-f293e1e4-71d8-454a-ad10-4ad3ace0dbff.png)

### PR Checklist
- [x] Does not introduce dependencies
- [x] **Functional:** all changes do not break existing APIs and if so, bump major version.
- [x] **Tests pass:** check the output of yarn test packages/<Kongponent>
- [x] **Naming:** the files and the method and prop variables use the same naming conventions as other Kongponents
- [x] **Framework style:** abides by the essential rules in Vue's style guide
- [x] **Cleanliness:** does not have formatting issues, unused code (e.g., console.logs), or leftover comments
- [x] **Docs:** includes a technically accurate README, uses JSDOC where appropriate

chrisvdp commented on packages/KInput/KInput.vue:
```diff
@@ -2,24 +2,35 @@
       class="col-md-4 mt-5">
       <div class="text-on-input">
-        <label :class="{ focused: isFocused }">
+        <label :class="{ focused: isFocused, hovered: isHovered }">
```
The label here is missing the `for` attr. I was playing with a library called `vue-unique-id` to handle cases like this. The other option, that I think also helps simplify the hover and focus styling is wrapping the label around the input. In this case, it doesn't require an id/for to associate the label and input.

kaiarrowood replied:
Updated it so if there is a label passed in we are now using `for` attribute.

chrisvdp commented on packages/KInput/KInput.vue:
```diff
@@ -2,24 +2,35 @@
   <div class="k-input-wrapper">
     <input
       v-if="!label"
```
every input should have a label. It can be visually hidden for style purposes

kaiarrowood replied:
Will not address as we doc correct usage of `KLabel` with `KInput`.

![image](https://user-images.githubusercontent.com/67973710/134971930-4084e16c-058f-4f20-9878-579e9d4db995.png)

chrisvdp commented on packages/KInput/KInput.vue:
```diff
@@ -49,11 +65,25 @@ export default {
+    /**
+     * Debug mode - for testing only, strips out generated ids
+     */
+    debugMode: {
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**nitpick:** ‏is there a better name than debug mode? Seems like something that should get turned off, maybe `testingMode`? or `staticIds`?

kaiarrowood replied:
Changed to `testMode`.

dustinryerson reviewed (dismissed):
Overall this is looking good! I have a few comments:
1. It looks like the hover and focus color changes are being applied via box-shadow rather than the border color. I think we should update it to modify the border color and remove the box-shadow. Below are some screenshots for reference:

**Kongponents**
Hover:
![Screen Shot 2021-09-28 at 11 16 00 AM](https://user-images.githubusercontent.com/2080476/135148544-248dd042-8d46-48c9-8b78-134c4f8a3f6f.png)
![hover](https://user-images.githubusercontent.com/2080476/135147765-556092e5-96f1-40b6-a40a-ab5a50fe2989.png)
Focus:
![Screen Shot 2021-09-28 at 11 15 33 AM](https://user-images.githubusercontent.com/2080476/135148385-a871db98-3dfe-48ed-aae8-4517c79c1a26.png)
![focus](https://user-images.githubusercontent.com/2080476/135147880-c56ce4ac-ebfd-4acf-8038-fa6166820351.png)
Focus (error):
![Screen Shot 2021-09-28 at 11 21 17 AM](https://user-images.githubusercontent.com/2080476/135148030-98429ccb-6341-4ad4-b8aa-1aa032cb79a9.png)

**Figma Design**
![Screen Shot 2021-09-28 at 11 18 08 AM](https://user-images.githubusercontent.com/2080476/135148124-32c26647-14ee-4c2d-8db3-296b0dae5d73.png)

2. This is a smaller nitpick - I think the input label should have a little more padding on the x-axis.
**Kongponent**
![Screen Shot 2021-09-28 at 11 56 48 AM](https://user-images.githubusercontent.com/2080476/135148928-e2b46f76-94be-457d-a7ba-820524f6e8d4.png)
**Figma**
![Screen Shot 2021-09-28 at 11 56 25 AM](https://user-images.githubusercontent.com/2080476/135148904-b165fc0e-c436-4910-98bd-0a6586ec2260.png)

3. I think it might improve the UX to add a slight transition when changing the color on hover and focus, that way the color change isn't so immediate (which I think some users may experience as a little jarring)

4. I may have just missed it, but I couldn't find any design mockup for the input field error state border color and label styling. If it doesn't exist, it would be helpful to get it added to Figma so that we can reference it during the review.

kaiarrowood commented:
Hey @dustinryerson , thanks for the review. I have addressed the label horizontal padding and fixed the error state display. I am using `box-shadow` instead of `border` so that I can use the `inset` property. `border` does not differentiate between if it is an inner border or an outer border - in the new designs the focus "border" needs a width of `2px`, so to avoid screen shift on focus/hover I am using an `inset` `box-shadow`.

I could add a transition, but I'm not familiar with that CSS prop; do you have a suggestion on what you think may be a good one?

Lastly, no, there isn't any Figma related to error state at this point.
